File: loan_manager/server/routers/customers.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import Customer, Order, RepaymentPlan, User
from ..schemas import CustomerCreate, CustomerUpdate, CustomerOut
from ..auth_deps import get_active_user

router = APIRouter(prefix="/api/customers", tags=["customers"])


def _next_customer_no(db: Session, tenant_id: int) -> int:
    """Get the next sequential customer_no for a tenant."""
    max_no = db.query(func.max(Customer.customer_no)).filter(
        Customer.tenant_id == tenant_id
    ).scalar()
    return (max_no or 0) + 1


@router.get("/", response_model=list[CustomerOut])
def list_customers(
    keyword: str = Query("", description="搜索关键词"),
    skip: int = 0,
    limit: int = 100,
    user: User = Depends(get_active_user),
    db: Session = Depends(get_db),
):
    q = db.query(Customer).filter(Customer.tenant_id == user.tenant_id)
    if keyword:
        q = q.filter(
            Customer.name.contains(keyword)
            | Customer.phone.contains(keyword)
            | Customer.id_card.contains(keyword)
        )
    return q.order_by(Customer.customer_no.asc()).offset(skip).limit(limit).all()


@router.get("/count")
def count_customers(keyword: str = "", user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    q = db.query(Customer).filter(Customer.tenant_id == user.tenant_id)
    if keyword:
        q = q.filter(
            Customer.name.contains(keyword)
            | Customer.phone.contains(keyword)
            | Customer.id_card.contains(keyword)
        )
    return {"count": q.count()}


@router.get("/{customer_id}", response_model=CustomerOut)
def get_customer(customer_id: int, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    c = db.query(Customer).filter(Customer.id == customer_id, Customer.tenant_id == user.tenant_id).first()
    if not c:
        raise HTTPException(status_code=404, detail="客户不存在")
    return c


@router.post("/", response_model=CustomerOut)
def create_customer(data: CustomerCreate, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    d = data.model_dump()
    # If id_card provided, check for existing customer with same id_card (auto-merge)
    if d.get("id_card"):
        existing_by_id_card = db.query(Customer).filter(
            Customer.tenant_id == user.tenant_id,
            Customer.id_card == d["id_card"],
        ).first()
        if existing_by_id_card:
            # Update non-empty fields and return existing record
            for k, v in d.items():
                if k not in ("customer_no",) and v not in (None, "", 0):
                    setattr(existing_by_id_card, k, v)
            db.commit()
            db.refresh(existing_by_id_card)
            return existing_by_id_card
    # Auto-assign customer_no if not provided or zero
    if not d.get("customer_no"):
        d["customer_no"] = _next_customer_no(db, user.tenant_id)
    else:
        existing = db.query(Customer).filter(
            Customer.tenant_id == user.tenant_id,
            Customer.customer_no == d["customer_no"],
        ).first()
        if existing:
            raise HTTPException(status_code=400, detail=f"编号 {d['customer_no']} 已被使用")
    c = Customer(**d, tenant_id=user.tenant_id)
    db.add(c)
    db.commit()
    db.refresh(c)
    return c


@router.post("/merge-duplicates")
def merge_duplicate_customers(user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    """Find all customers with same id_card in the tenant, merge duplicates into the oldest record."""
    from ..models import Appointment
    # Find id_cards that appear more than once
    dup_query = (
        db.query(Customer.id_card, func.count(Customer.id).label("cnt"))
        .filter(Customer.tenant_id == user.tenant_id, Customer.id_card != "")
        .group_by(Customer.id_card)
        .having(func.count(Customer.id) > 1)
        .all()
    )
    merged_groups = 0
    merged_customers = 0
    for id_card, _ in dup_query:
        group = (
            db.query(Customer)
            .filter(Customer.tenant_id == user.tenant_id, Customer.id_card == id_card)
            .order_by(Customer.id.asc())
            .all()
        )
        if len(group) < 2:
            continue
        primary = group[0]
        # Merge info from duplicates into primary (fill empty fields)
        for dup in group[1:]:
            for col in ("name", "phone", "address", "email", "account_manager", "emergency_contact"):
                if not getattr(primary, col) and getattr(dup, col):
                    setattr(primary, col, getattr(dup, col))
            # Reassign orders and appointments to primary
            db.query(Order).filter(Order.customer_id == dup.id).update({"customer_id": primary.id})
            db.query(Appointment).filter(Appointment.customer_id == dup.id).update({"customer_id": primary.id})
            db.flush()
            db.delete(dup)
            merged_customers += 1
        merged_groups += 1
    db.commit()
    return {"merged_groups": merged_groups, "removed_duplicates": merged_customers}


@router.put("/{customer_id}", response_model=CustomerOut)
def update_customer(customer_id: int, data: CustomerUpdate, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    c = db.query(Customer).filter(Customer.id == customer_id, Customer.tenant_id == user.tenant_id).first()
    if not c:
        raise HTTPException(status_code=404, detail="客户不存在")
    updates = data.model_dump(exclude_unset=True)
    if "customer_no" in updates and updates["customer_no"]:
        existing = db.query(Customer).filter(
            Customer.tenant_id == user.tenant_id,
            Customer.customer_no == updates["customer_no"],
            Customer.id != customer_id,
        ).first()
        if existing:
            raise HTTPException(status_code=400, detail=f"编号 {updates['customer_no']} 已被使用")
    for k, v in updates.items():
        setattr(c, k, v)
    db.commit()
    db.refresh(c)
    return c


@router.delete("/{customer_id}")
def delete_customer(customer_id: int, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    c = db.query(Customer).filter(Customer.id == customer_id, Customer.tenant_id == user.tenant_id).first()
    if not c:
        raise HTTPException(status_code=404, detail="客户不存在")
    db.delete(c)
    db.commit()
    return {"message": "删除成功"}


@router.get("/{customer_id}/overview")
def customer_overview(customer_id: int, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    """Return customer info + order history with repayment summary."""
    c = db.query(Customer).filter(Customer.id == customer_id, Customer.tenant_id == user.tenant_id).first()
    if not c:
        raise HTTPException(status_code=404, detail="客户不存在")

    orders = db.query(Order).filter(Order.customer_id == customer_id, Order.tenant_id == user.tenant_id).order_by(Order.id).all()
    order_list = []
    for o in orders:
        total_price = float((o.unit_price + o.processing_fee) * o.weight)
        down_payment = float(o.down_payment)
        notary_fee = float(o.notary_fee)

        plans = db.query(RepaymentPlan).filter(
            RepaymentPlan.order_id == o.id
        ).order_by(RepaymentPlan.period_no).all()

        total_paid = sum(float(p.paid_amount) for p in plans)
        balance = total_price - down_payment - total_paid

        settlement_date = None
        settlement_days = None
        has_overdue = any("逾期" in (p.status or "") for p in plans)
        # Check if all periods are fully paid (actual settlement)
        all_paid = plans and all(
            float(p.paid_amount) >= float(p.total_amount) and p.paid_date
            for p in plans
        )
        if all_paid or o.status == "已结清":
            display_status = "已结清"
            paid_dates = [p.paid_date for p in plans if p.paid_date]
            if paid_dates:
                last_paid = max(paid_dates)
                settlement_date = str(last_paid)
                settlement_days = (last_paid - o.order_date).days
        elif has_overdue:
            display_status = "已逾期"
        else:
            display_status = "正常"

        order_list.append({
            "order_no": o.order_no,
            "weight": float(o.weight),
            "total_price": round(total_price),
            "down_payment": round(down_payment),
            "notary_fee": round(notary_fee),
            "installment_amount": float(o.installment_amount),
            "installment_periods": o.installment_periods,
            "order_date": str(o.order_date),
            "status": display_status,
            "settlement_date": settlement_date,
            "settlement_days": settlement_days,
            "balance": round(balance, 2),
        })

    return {
        "customer": {
            "id": c.id,
            "customer_no": c.customer_no,
            "name": c.name,
            "phone": c.phone,
            "id_card": c.id_card,
            "address": c.address,
            "email": c.email,
            "account_manager": c.account_manager,
            "emergency_contact": c.emergency_contact,
            "has_overdue": c.has_overdue,
            "has_property": c.has_property,
        },
        "orders": order_list,
    }

File: loan_manager/loan-manager-cf/src/routes/customers.ts
import { Hono } from 'hono';
import { drizzle } from 'drizzle-orm/d1';
import { eq, and, like, or, sql, max, desc, asc, inArray } from 'drizzle-orm';
import { customers, orders, repaymentPlans, appointments } from '../db/schema';
import { authMiddleware, writeProtectMiddleware } from '../middleware/auth';
import type { Env } from '../lib/types';

const customersRoute = new Hono<Env>();
customersRoute.use('*', authMiddleware, writeProtectMiddleware);

function customerOut(c: any) {
  return {
    id: c.id,
    tenant_id: c.tenantId,
    customer_no: c.customerNo,
    name: c.name,
    phone: c.phone || '',
    id_card: c.idCard || '',
    address: c.address || '',
    email: c.email || '',
    account_manager: c.accountManager || '',
    emergency_contact: c.emergencyContact || '',
    has_overdue: c.hasOverdue || '否',
    has_property: c.hasProperty || '否',
    created_at: c.createdAt || '',
  };
}

// ── List ─────────────────────────────────────────────────
customersRoute.get('/', async (c) => {
  const user = c.get('user');
  const keyword = c.req.query('keyword') || '';
  const skip = parseInt(c.req.query('skip') || '0');
  const limit = parseInt(c.req.query('limit') || '100');
  const db = drizzle(c.env.DB);

  let q = db.select().from(customers).where(eq(customers.tenantId, user.tenantId!));
  if (keyword) {
    q = db.select().from(customers).where(and(
      eq(customers.tenantId, user.tenantId!),
      or(
        like(customers.name, `%${keyword}%`),
        like(customers.phone, `%${keyword}%`),
        like(customers.idCard, `%${keyword}%`),
      ),
    ));
  }
  const rows = await q.orderBy(asc(customers.customerNo)).offset(skip).limit(limit);
  return c.json(rows.map(customerOut));
});

// ── Count ────────────────────────────────────────────────
customersRoute.get('/count', async (c) => {
  const user = c.get('user');
  const keyword = c.req.query('keyword') || '';
  const db = drizzle(c.env.DB);

  let where = eq(customers.tenantId, user.tenantId!);
  if (keyword) {
    where = and(where, or(
      like(customers.name, `%${keyword}%`),
      like(customers.phone, `%${keyword}%`),
      like(customers.idCard, `%${keyword}%`),
    ))!;
  }
  const result = await db.select({ count: sql<number>`count(*)` }).from(customers).where(where);
  return c.json({ count: result[0]?.count || 0 });
});

// ── Merge duplicates ─────────────────────────────────────
customersRoute.post('/merge-duplicates', async (c) => {
  const user = c.get('user');
  const db = drizzle(c.env.DB);
  const tid = user.tenantId!;

  // Find id_cards with duplicates
  const dups = await db.select({
    idCard: customers.idCard,
    cnt: sql<number>`count(*)`,
  }).from(customers)
    .where(and(eq(customers.tenantId, tid), sql`${customers.idCard} != ''`))
    .groupBy(customers.idCard)
    .having(sql`count(*) > 1`);

  let mergedGroups = 0;
  let removedDuplicates = 0;

  for (const dup of dups) {
    const group = await db.select().from(customers)
      .where(and(eq(customers.tenantId, tid), eq(customers.idCard, dup.idCard!)))
      .orderBy(asc(customers.id));
    if (group.length < 2) continue;

    const primary = group[0];
    for (let i = 1; i < group.length; i++) {
      const d = group[i];
      // Fill empty fields from duplicate into primary
      const updates: Record<string, any> = {};
      for (const col of ['name', 'phone', 'address', 'email', 'accountManager', 'emergencyContact'] as const) {
        if (!primary[col] && d[col]) updates[col] = d[col];
      }
      if (Object.keys(updates).length > 0) {
        await db.update(customers).set(updates).where(eq(customers.id, primary.id));
      }
      // Reassign orders and appointments
      await db.update(orders).set({ customerId: primary.id }).where(eq(orders.customerId, d.id));
      await db.update(appointments).set({ customerId: primary.id }).where(eq(appointments.customerId, d.id));
      // Delete duplicate
      await db.delete(customers).where(eq(customers.id, d.id));
      removedDuplicates++;
    }
    mergedGroups++;
  }
  return c.json({ merged_groups: mergedGroups, removed_duplicates: removedDuplicates });
});

// ── Get by ID ────────────────────────────────────────────
customersRoute.get('/:id', async (c) => {
  const user = c.get('user');
  const id = parseInt(c.req.param('id'));
  if (isNaN(id)) return c.json({ detail: '无效的ID' }, 400);

  // Check if this is the overview endpoint
  const db = drizzle(c.env.DB);
  const cust = await db.select().from(customers)
    .where(and(eq(customers.id, id), eq(customers.tenantId, user.tenantId!))).get();
  if (!cust) return c.json({ detail: '客户不存在' }, 404);
  return c.json(customerOut(cust));
});

// ── Customer Overview ────────────────────────────────────
customersRoute.get('/:id/overview', async (c) => {
  const user = c.get('user');
  const id = parseInt(c.req.param('id'));
  const db = drizzle(c.env.DB);

  const cust = await db.select().from(customers)
    .where(and(eq(customers.id, id), eq(customers.tenantId, user.tenantId!))).get();
  if (!cust) return c.json({ detail: '客户不存在' }, 404);

  const orderList = await db.select().from(orders)
    .where(and(eq(orders.customerId, id), eq(orders.tenantId, user.tenantId!)))
    .orderBy(asc(orders.id));

  const orderResults = [];
  for (const o of orderList) {
    const totalPrice = (parseFloat(o.unitPrice || '0') + parseFloat(o.processingFee || '0')) * parseFloat(o.weight || '0');
    const downPayment = parseFloat(o.downPayment || '0');

    const plans = await db.select().from(repaymentPlans)
      .where(eq(repaymentPlans.orderId, o.id))
      .orderBy(asc(repaymentPlans.periodNo));

    const totalPaid = plans.reduce((s, p) => s + parseFloat(p.paidAmount || '0'), 0);
    const balance = totalPrice - downPayment - totalPaid;

    const hasOverdue = plans.some(p => (p.status || '').includes('逾期'));
    const allPaid = plans.length > 0 && plans.every(p =>
      parseFloat(p.paidAmount || '0') >= parseFloat(p.totalAmount || '0') && p.paidDate
    );

    let displayStatus = '正常';
    let settlementDate: string | null = null;
    let settlementDays: number | null = null;

    if (allPaid || o.status === '已结清') {
      displayStatus = '已结清';
      const paidDates = plans.filter(p => p.paidDate).map(p => p.paidDate!);
      if (paidDates.length > 0) {
        const lastPaid = paidDates.sort().reverse()[0];
        settlementDate = lastPaid;
        const orderDt = new Date(o.orderDate);
        const lastDt = new Date(lastPaid);
        settlementDays = Math.round((lastDt.getTime() - orderDt.getTime()) / 86400000);
      }
    } else if (hasOverdue) {
      displayStatus = '已逾期';
    }

    orderResults.push({
      order_no: o.orderNo,
      weight: parseFloat(o.weight || '0'),
      total_price: Math.round(totalPrice),
      down_payment: Math.round(downPayment),
      notary_fee: Math.round(parseFloat(o.notaryFee || '0')),
      installment_amount: parseFloat(o.installmentAmount || '0'),
      installment_periods: o.installmentPeriods || 0,
      order_date: o.orderDate,
      status: displayStatus,
      settlement_date: settlementDate,
      settlement_days: settlementDays,
      balance: Math.round(balance * 100) / 100,
    });
  }

  return c.json({
    customer: {
      id: cust.id,
      customer_no: cust.customerNo,
      name: cust.name,
      phone: cust.phone || '',
      id_card: cust.idCard || '',
      address: cust.address || '',
      email: cust.email || '',
      account_manager: cust.accountManager || '',
      emergency_contact: cust.emergencyContact || '',
      has_overdue: cust.hasOverdue || '否',
      has_property: cust.hasProperty || '否',
    },
    orders: orderResults,
  });
});

// ── Create ───────────────────────────────────────────────
customersRoute.post('/', async (c) => {
  const user = c.get('user');
  const body = await c.req.json();
  const db = drizzle(c.env.DB);
  const tid = user.tenantId!;

  // Auto-merge by id_card
  if (body.id_card) {
    const existing = await db.select().from(customers)
      .where(and(eq(customers.tenantId, tid), eq(customers.idCard, body.id_card))).get();
    if (existing) {
      const updates: Record<string, any> = {};
      if (body.name) updates.name = body.name;
      if (body.phone) updates.phone = body.phone;
      if (body.address) updates.address = body.address;
      if (body.email) updates.email = body.email;
      if (body.account_manager) updates.accountManager = body.account_manager;
      if (body.emergency_contact) updates.emergencyContact = body.emergency_contact;
      if (body.has_overdue) updates.hasOverdue = body.has_overdue;
      if (body.has_property) updates.hasProperty = body.has_property;
      if (Object.keys(updates).length > 0) {
        await db.update(customers).set(updates).where(eq(customers.id, existing.id));
      }
      const updated = await db.select().from(customers).where(eq(customers.id, existing.id)).get();
      return c.json(customerOut(updated));
    }
  }

  // Auto-assign customer_no
  let customerNo = body.customer_no || 0;
  if (!customerNo) {
    const maxResult = await db.select({ maxNo: max(customers.customerNo) }).from(customers)
      .where(eq(customers.tenantId, tid));
    customerNo = (maxResult[0]?.maxNo || 0) + 1;
  } else {
    const existing = await db.select().from(customers)
      .where(and(eq(customers.tenantId, tid), eq(customers.customerNo, customerNo))).get();
    if (existing) return c.json({ detail: `编号 ${customerNo} 已被使用` }, 400);
  }

  const result = await db.insert(customers).values({
    tenantId: tid,
    customerNo,
    name: body.name,
    phone: body.phone || '',
    idCard: body.id_card || '',
    address: body.address || '',
    email: body.email || '',
    accountManager: body.account_manager || '',
    emergencyContact: body.emergency_contact || '',
    hasOverdue: body.has_overdue || '否',
    hasProperty: body.has_property || '否',
  }).returning();
  return c.json(customerOut(result[0]));
});

// ── Update ───────────────────────────────────────────────
customersRoute.put('/:id', async (c) => {
  const user = c.get('user');
  const id = parseInt(c.req.param('id'));
  const body = await c.req.json();
  const db = drizzle(c.env.DB);

  const cust = await db.select().from(customers)
    .where(and(eq(customers.id, id), eq(customers.tenantId, user.tenantId!))).get();
  if (!cust) return c.json({ detail: '客户不存在' }, 404);

  if (body.customer_no && body.customer_no !== cust.customerNo) {
    const dup = await db.select().from(customers)
      .where(and(
        eq(customers.tenantId, user.tenantId!),
        eq(customers.customerNo, body.customer_no),
        sql`${customers.id} != ${id}`,
      )).get();
    if (dup) return c.json({ detail: `编号 ${body.customer_no} 已被使用` }, 400);
  }

  const updates: Record<string, any> = {};
  if (body.customer_no !== undefined) updates.customerNo = body.customer_no;
  if (body.name !== undefined) updates.name = body.name;
  if (body.phone !== undefined) updates.phone = body.phone;
  if (body.id_card !== undefined) updates.idCard = body.id_card;
  if (body.address !== undefined) updates.address = body.address;
  if (body.email !== undefined) updates.email = body.email;
  if (body.account_manager !== undefined) updates.accountManager = body.account_manager;
  if (body.emergency_contact !== undefined) updates.emergencyContact = body.emergency_contact;
  if (body.has_overdue !== undefined) updates.hasOverdue = body.has_overdue;
  if (body.has_property !== undefined) updates.hasProperty = body.has_property;

  await db.update(customers).set(updates).where(eq(customers.id, id));
  const updated = await db.select().from(customers).where(eq(customers.id, id)).get();
  return c.json(customerOut(updated));
});

// ── Delete ───────────────────────────────────────────────
customersRoute.delete('/:id', async (c) => {
  const user = c.get('user');
  const id = parseInt(c.req.param('id'));
  const db = drizzle(c.env.DB);

  const cust = await db.select().from(customers)
    .where(and(eq(customers.id, id), eq(customers.tenantId, user.tenantId!))).get();
  if (!cust) return c.json({ detail: '客户不存在' }, 404);

  await db.delete(customers).where(eq(customers.id, id));
  return c.json({ message: '删除成功' });
});

export default customersRoute;
